core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1963 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use self::{
+    HandleType::{
+        Internal,
+        Leaf,
+    },
+    InsertResult::{
+        Fit,
+        Split,
+    },
+};
+use super::search::{
+    self,
+    SearchResult::{
+        Found,
+        NotFound,
+    },
+};
+use crate::storage::{
+    self,
+    alloc::{
+        Allocate,
+        AllocateUsing,
+        Initialize,
+    },
+    btree_map::node::{
+        KVHandle,
+        KVPair,
+        KVRef,
+        Node,
+        NodeHandle,
+    },
+    chunk::SyncChunk,
+    Flush,
+};
+use core::{
+    borrow::Borrow,
+    cmp::Ord,
+};
+#[cfg(feature = "ink-generate-abi")]
+use ink_abi::{
+    HasLayout,
+    LayoutField,
+    LayoutStruct,
+    StorageLayout,
+};
+use scale::{
+    Codec,
+    Decode,
+    Encode,
+};
+#[cfg(feature = "ink-generate-abi")]
+use type_metadata::Metadata;
+
+/// Each node in the tree has 2 * B children.
+///
+/// For the moment we use the same number as the Rust standard library
+/// implementation here. There is no empiric evidence for the effectiveness
+/// of this number yet -- we just chose this number to have a first starting
+/// point.
+pub(super) const B: usize = 6;
+
+/// Number of elements (i.e. key/value pairs) which can be stored in one
+/// node of the tree. The `- 1` is because there needs to be an edge to
+/// the right of the last element.
+///
+/// ```no_compile
+/// const B: usize = 2;
+/// const CAPACITY: usize = 2 * B - 1;
+/// keys  = [    a,    b,    c    ];
+/// edges = [ 1,    2,    3,    4 ];
+/// ```
+pub(super) const CAPACITY: usize = 2 * B - 1;
+
+/// Pointer to the index where a key/value pair is stored in `kv_pairs`.
+pub(super) type KVStoragePointer = u32;
+
+/// The node type, either a `Leaf` (a node without children) or `Internal`
+/// (a node with children). This distinction makes it easier to handle e.g.
+/// cases where nodes are merged or implementing recursion.
+pub(super) enum HandleType {
+    Leaf,
+    Internal,
```

Comment:
I agree, `Branch` seems a term which is a lot more speaking. I originally took the naming from the Rust implementation. I renamed it now everywhere. `Internal` was also confusing since our implementation also has an `InternalEntry` which is something completely different.